crates/router/src/core/payments/tokenization.rs

```diff
@@ -31,102 +31,115 @@ pub async fn save_payment_method<F: Clone, FData>(
                     };
-                }
-                Err(error) => {
-                    match error.current_context() {
-                        errors::StorageError::DatabaseError(err) => match err.current_context() {
-                            storage_models::errors::DatabaseError::NotFound => {
-                                let pm_metadata =
-                                    create_payment_method_metadata(None, connector_token)?;
-                                payment_methods::cards::create_payment_method(
-                                    db,
-                                    &payment_method_create_request,
-                                    &customer.customer_id,
-                                    &locker_response.0.payment_method_id,
-                                    merchant_id,
-                                    pm_metadata,
-                                )
-                                .await
-                                .change_context(errors::ApiErrorResponse::InternalServerError)
-                                .attach_printable("Failed to add payment method in db")
-                            }
-                            _ => Err(report!(errors::ApiErrorResponse::InternalServerError)),
-                        },
-                        _ => Err(report!(errors::ApiErrorResponse::InternalServerError)),
-                    }?;
-                }
+                } else {
+                    let pm_metadata = create_payment_method_metadata(None, connector_token)?;
+                    payment_methods::cards::create_payment_method(
+                        db,
+                        &payment_method_create_request,
+                        &customer.customer_id,
+                        &locker_response.0.payment_method_id,
+                        merchant_id,
+                        pm_metadata,
+                    )
+                    .await
+                    .change_context(errors::ApiErrorResponse::InternalServerError)
+                    .attach_printable("Failed to add payment method in db")?;
+                };
+                Some(locker_response.0.payment_method_id)
+            } else {
+                None
             };
-        } else {
-            let pm_metadata = create_payment_method_metadata(None, connector_token)?;
-            payment_methods::cards::create_payment_method(
-                db,
-                &payment_method_create_request,
-                &customer.customer_id,
-                &locker_response.0.payment_method_id,
-                merchant_id,
-                pm_metadata,
-            )
-            .await
-            .change_context(errors::ApiErrorResponse::InternalServerError)
-            .attach_printable("Failed to add payment method in db")?;
-        };
-        Some(locker_response.0.payment_method_id)
-    } else {
-        None
-    };
-    Ok(pm_id)
+            Ok(pm_id)
+        }
+        Err(_) => Ok(None),
```

Comment:
What are the standards followed on this,
do we not save the card because it might be invalid?

What if we receive errors that point to other errors such as insufficient balance etc...

Reply:
Connectors have different errors for card specific errors it's not easy enough to map them into one

Reply:
#1599 Added an issue to pick this up later